tensorflow/core/kernels/mkl_maxpooling_op.cc

```diff
@@ -525,64 +525,94 @@ class MklMaxPoolingOp : public MklPoolingForwardOpBase<T> {
-                                 this->data_format_mkldnn_, &output_tensor);
+      // Get the input memory descriptor
+      memory::desc input_md = dnn_shape_input.IsMklTensor()
+                                ? dnn_shape_input.GetMklLayout()
+                                : memory::desc(TFShapeToMklDnnDimsInNCHW(
+                                                 input_tensor_shape,
+                                                 this->data_format_tf_),
+                                          MklDnnType<T>(),
+                                          this->data_format_mkldnn_);
+
+      // Get src/filter/stride/padding information
+      memory::dims src_dims = dnn_shape_input.IsMklTensor()
+                            ? dnn_shape_input.GetSizesAsMklDnnDims()
+                            : TFShapeToMklDnnDimsInNCHW(input_tensor.shape(),
+                                  this->data_format_tf_);
+
+      memory::dims filter_dims, strides, padding_left, padding_right;
+      this->PoolParamsToDims(&pool_params, &filter_dims, &strides,
+          &padding_left, &padding_right);
+
+      // Get a pooling op from the cached pool
+      MklPoolingFwdPrimitive<T> *pooling_fwd = nullptr;
+      MklPoolingParams fwdParams(src_dims, output_dims_mkl_order, filter_dims,
+          strides, padding_left, padding_right, algorithm::pooling_max);
+      pooling_fwd = MklPoolingFwdPrimitiveFactory<T>::Get(fwdParams);
+
+      // allocate output tensor
+      this->AllocateOutputTensor(context, *(pooling_fwd->GetPoolingFwdPd()),
+          output_dims_mkl_order, this->data_format_mkldnn_, &output_tensor);
       OP_REQUIRES_OK(context, context->status());
-      dnn_data_output.SetUsrMemDataHandle(output_tensor);
+      dnn_data_output.SetUsrMem(output_dims_mkl_order,
+          pooling_fwd->GetDstMemoryFormat(), output_tensor);
 
-      AllocateWorkspaceTensor(context, pool_fwd_desc, &dnn_data_wksp);
+      AllocateWorkspaceTensor(context, *(pooling_fwd->GetPoolingFwdPd()),
+          &dnn_data_wksp);
       OP_REQUIRES_OK(context, context->status());
 
-      this->PrepareAndExecuteNet(pool_fwd_desc, &dnn_data_input,
-                                 &dnn_data_output, &dnn_data_wksp);
+      // check wehther we need to reorder src
+      T* src_data = nullptr;
+      if (input_md.data.format != pooling_fwd->GetSrcMemoryFormat()) {
+        dnn_data_input.SetUsrMem(input_md, &input_tensor);
+        auto src_target_primitive_desc = memory::primitive_desc(
+            {{src_dims}, MklDnnType<T>(), pooling_fwd->GetSrcMemoryFormat()},
+            cpu_engine);
+        dnn_data_input.CheckReorderToOpMem(src_target_primitive_desc);
+        src_data = static_cast<T*>(
+                    dnn_data_input.GetOpMem().get_data_handle());
+      } else {
+        src_data = static_cast<T*>(const_cast<T*>(
+                    input_tensor.flat<T>().data()));
+      }
+
+      T* dst_data = static_cast<T*>(
```

Comment:
I don't think you need any casts here. The underlying Eigen::TensorMap has overload of data() for for const and non-const. Same everywhere.

Reply:
Done with code change. No cast is needed